components/tools/OmeroPy/src/omero/plugins/metadata.py

```diff
@@ -0,0 +1,507 @@
+        else:
+            populate_metadata.log.setLevel(logging.INFO)
+
+        context_class = dict(self.POPULATE_CONTEXTS)[args.context]
+
+        fileid = args.fileid
+        cfgid = args.cfgid
+
+        if args.attach and not args.dry_run:
+            if args.file:
+                fileann = conn.createFileAnnfromLocalFile(
+                    args.file, mimetype=guess_mimetype(args.file),
+                    ns=NSBULKANNOTATIONSRAW)
+                fileid = fileann.getFile().getId()
+                md.linkAnnotation(fileann)
+
+            if args.cfg:
+                cfgann = conn.createFileAnnfromLocalFile(
+                    args.cfg, mimetype=guess_mimetype(args.cfg),
+                    ns=NSBULKANNOTATIONSCONFIG)
+                cfgid = cfgann.getFile().getId()
+                md.linkAnnotation(cfgann)
+
+        # Note some contexts only support a subset of these args
+        ctx = context_class(client, args.obj, file=args.file, fileid=fileid,
+                            cfg=args.cfg, cfgid=cfgid, attach=args.attach)
+        ctx.parse()
+        if not args.dry_run:
+            ctx.write_to_omero()
+
+    def rois(self, args):
+        "Manage ROIs"
+        md = self._load(args)
+        if args.delete:
+            graphspec = "/%s/Roi:%d" % (md.get_type(), md.get_id())
+            cmd = ["delete", graphspec]
+            if args.report:
+                cmd += ["--report"]
+            if args.dry_run:
+                cmd += ["--dry-run"]
+            return self.ctx.invoke(cmd)
+
+        client = self.ctx.conn(args)
+        params = omero.sys.ParametersI()
+        params.addId(md.get_id())
+        if md.get_type() == "Screen":
+            q = """SELECT r.id FROM Roi r, WellSample ws, Plate p,
+                   ScreenPlateLink spl WHERE
+                   spl.child=p AND r.image=ws.image AND ws.well.plate=p AND
+                   spl.parent.id=:id"""
+        elif md.get_type() == "Plate":
+            q = """SELECT r.id FROM Roi r, WellSample ws WHERE
+                   r.image=ws.image AND ws.well.plate.id=:id"""
+        elif md.get_type() == "PlateAcquisition":
+            q = """SELECT r.id FROM Roi r, WellSample ws WHERE
+                   r.image=ws.image AND ws.plateAcquisition.id=:id"""
+        elif md.get_type() == "Well":
+            q = """SELECT r.id FROM Roi r, WellSample ws WHERE
+                   r.image=ws.image AND ws.well.id=:id"""
+        elif md.get_type() == "Project":
+            q = """SELECT r.id FROM Roi r, DatasetImageLink dil,
+                   ProjectDatasetLink pdl WHERE dil.child=r.image AND
+                   dil.parent=pdl.child AND pdl.parent.id=:id"""
+        elif md.get_type() == "Dataset":
+            q = """SELECT r.id FROM Roi r, DatasetImageLink dil WHERE
+                   dil.child=r.image AND dil.parent.id=:id"""
+        elif md.get_type() == "Image":
+            q = """SELECT r.id FROM Roi r WHERE r.image.id=:id"""
```

Comment:
This was one of the reasons I added query-by-graph-spec as an RFE, since this could be reduced to
```
queryByGraphSpec(md.get_type() + '/*/Roi')
```